feat(compiler): support expressions in list and object index

adnrs96 opened the pull request:
**- What I did**
In this commit we start supporting expressions in the list or
object index.

**- How I did it**
We do this by modifying grammar to accept expressions as list or
an object index and work on path resolution code path to interpret
these expressions and spit out correct resolved types.

**- How to verify it**
This means following syntax is now supported:
Examples:
		a[2 + 2]
		a["foo" + "bar"]

Closes: #559.

adnrs96 commented on storyscript/parser/Tree.py:
```diff
@@ -120,6 +120,23 @@ def position(self):
         """
         return Position(self.line(), self.column(), self.end_column())
 
+    def collect_expression(self):
```
Is there anything better available that could help provide me with the syntax that was parsed in the first place to generate a particular sub Tree?
I looked around and didn't find anything in the code base. Something similar is the extract_path function I guess.
Lark losses the content that was parsed. I would have imagined it storing the raw code in every tree's node which is basically what a particular tree node and all its children represent altogether.

Anyway, I wrote this function to handle the square bracket insertion because that's what I thought of as the possible use case inside an expression (eg. `b[b["foo"]]`).

wilzbach replied:
Yes, there's: https://github.com/storyscript/storyscript/pull/775
We need to revive this one.

adnrs96 commented on tests/e2e/object_int_index_invalid2.error:
```diff
@@ -3,4 +3,4 @@ Error: syntax error in story at line 2, column 1
       ^^^^^^
 
-E0104: `Map[int,boolean]` can't be indexed with `-2` of type `string`
+E0104: `Map[int,boolean]` can't be indexed with `"-2"` of type `string`
```
We end up inserting these brackets around `"-2"` which look weird at first but in case of an expression like `"2" + "2"`, I believe it would make sense. In general as well it makes it more clear I guess since `-2` was a string in the first place.

wilzbach replied:
Well, the old formatting distinguished between strings and numbers and there's no reason not to.

wilzbach commented on storyscript/compiler/semantics/ExpressionResolver.py:
```diff
@@ -189,7 +189,10 @@ class ExpressionResolver:
-        self.path_resolver = PathResolver(symbol_resolver=symbol_resolver)
+        self.path_resolver = PathResolver(
+            symbol_resolver=symbol_resolver,
+            expression_resolver=self,
```
The idea here was to encapsulate responsibilities and making the `PathResolver` as dump as possible.

wilzbach commented on storyscript/compiler/semantics/PathResolver.py:
```diff
@@ -37,22 +29,20 @@ def names(self, tree):
+                if child.data == 'expression':
+                    type_ = self.resolver.expression(child).type()
+                    value = child.collect_expression()
+                    if '__p-' in value:
```
There should be a `is_internal` method.

wilzbach commented on storyscript/compiler/json/Objects.py:
```diff
@@ -22,15 +22,11 @@ def names(self, tree):
-                elif child.data == 'number':
-                    value = self.number(child)
+                if child.data == 'expression':
+                    value = self.expression(child)
```
So here's the thing. We would broke the JSON protocol with this change. You have two options:
a) Make the change in the platform-engine + make sure the testsuite still passes
b) Lowering expressions into fake lines instead

adnrs96 replied:
I have been reading up on conversation that happened over here [#593](https://github.com/storyscript/storyscript/issues/593) and it feels like going with the second option might make things tricky since we also want to introduce the ternary operators [#422](https://github.com/storyscript/storyscript/issues/422)

I can definitely investigate platform-engine code and see how to modify things over there to maintain the parity of protocol.

Thoughts?

judepereira replied:
I like the direction on updating the runtime.

judepereira replied:
What's the overhead in attempting to lower it?

wilzbach replied:
> second option might make things tricky since we also want to introduce the ternary operators #422

How so?

> What's the overhead in attempting to lower it?

I don't see any.

adnrs96 replied:
I was worried about cases where this kinda lowering would be nested basically and might interact with other types of lowering phases resulting in a dead lock kinda situation.

I gave it a bit of thought and went ahead with the lowering approach.

adnrs96 commented:
Hey @wilzbach @judepereira 
Thanks for taking a look and suggesting change in approach.

I have updated this PR with the lowering approach in mind. Took quite a time due to subtle issue with the line numbers. Basically somewhere upward in another lowering we were using same token terminal in two places and when I did the lowering step in respect to path, that single terminal in two places kept messing up the line number update and ended up causing not unquie line error!

It would be great if you guys can take a look and review. Looking forward to your thoughts :)

PS: I don't know why the coverage is dropping in this PR by 2%.

wilzbach commented on storyscript/compiler/lowering/Lowering.py:
```diff
@@ -584,6 +605,45 @@ def visit_function_dot(self, node, block):
+            block = self.fake_tree(node)
+        elif node.data == 'path':
+            for child in node.children:
+                if not isinstance(child, Tree):
```
```suggestion
                if not isinstance(child, Tree) or child.data != 'path_fragment':
```

then you get rid of one level of indentation.

adnrs96 commented on storyscript/compiler/lowering/Lowering.py:
```diff
@@ -584,6 +605,45 @@ def visit_function_dot(self, node, block):
+                        fake_path = block.add_assignment(
+                            expression,
+                            original_line=node.line())
+                        path_fragment.children = [fake_path]
```
I assume by overhead you meant ends up adding lot of fake lines even in cases where they aren't required. Eg. Case of single elements in the the expression.
```
a = [1, 2, 3]
a[1]
```

Now since you have mentioned it what I think could be done is, maybe selectively lower either into new lines or alter in place into basic string or number type. We already support the JSON on the platform which understands a `number` or `string` inside those parenthesis so we should be good.

wilzbach replied:
> I assume by overhead you meant ends up adding lot of fake lines even in cases where they aren't required. 

Yes.

> maybe selectively lower either into new lines or alter in place into basic string or number type.

Yes, my point.

adnrs96 commented:
Fixing destructuring totally makes sense. I thought there was some kinda reason behind why fake lines were inserted in a different way for destructuring and hence tried to go along and put out the fires it was causing.

adnrs96 commented:
I am not so sure what are you trying to say in the second point but I am sure we are talking about the same page with the `mark_lines` function. I don't get whether you are suggesting to make copies elsewhere (like when I move things to other block) or are you asking why the copies are being made in the first place for somecases.

Regardless I believe to fix the account for the third bullet point I might be able to avoid fiddling with moving things around in the second one all together so lets see how that turns out.

adnrs96 commented:
Yeah some sorta pretty printing code would be awesome over there instead of just saying `expression`.
I thought we would add it up later on fix this but since you have mentioned it, let me study out the your draft PR and see how I can use it here or whether we can add AST pretty printing first before going forward with this PR.

adnrs96 commented on storyscript/compiler/lowering/Lowering.py:
```diff
@@ -450,13 +450,34 @@ def visit_assignment(self, node, block, parent):
+                    )
                     name.line = new_line  # update token's line info
+                    # Update the line number for the original path token
+                    n.child(0).line = new_line
```
So what I had in mind is this. At any point in time, each token inside the AST of a particular line should have the same line number. Though we use an algorithm to detect line number of an AST much simpler than that by simply finding the first token and getting its line number.

I wanted to make line number's coherant even though in a future commit I am going to change it in the next visit_path setup. I believe this helps in cases where we make changes somewhere down the line and getting stuck becase the line numbers aren't behaving as they are expected to be. I mean aren't same inside the AST of a single line.

wilzbach replied:
> I believe this helps in cases where we make changes somewhere down the line and getting stuck becase the line numbers aren't behaving as they are expected to be.

Well, the thing is that `n.child(0)` gets removed from the AST later on with `node.children[0] = n` in L485 for the last element of the destucturing.

Or kept in the special case of (n == 1), but in either cases the update isn't necessary.
On the contrary with the current code, this might lead to the last line being incorrectly detected as fake line.

adnrs96 commented on storyscript/compiler/lowering/Lowering.py:
```diff
@@ -450,13 +450,34 @@ def visit_assignment(self, node, block, parent):
+                        orig_obj.child(0),
+                        'NAME',
+                        orig_obj.child(0).value
+                    )
```
So there are different path's and I believe I could have maybe done better with the naming.
```
{a} = {"a":20}
```
becomes
```
obj = {"a":20}
a = obj["a"]
```

Now we have an original copy of `obj` and we need to create a copy of it for each of the assignment lines we add like `a = obj["a"]` so that line number is different and coherant for each line.

Then there is token for `a` which is being used in two places, one for saying `path a` and the other one for saying `string a`. Thing is both of the token may look the same but need to be separate since if we move one of them into a different line and update its line number (`mark_line` does this) that change starts reflecting in the original line as well which was using the same token for `path a`.

This is why I had to create copies in that fashion. Each one of those copies are used in different places and to avoid issues when modifying AST these copies are needed.

I believe even though this ends up creating a lot of copies they should be done since essentially those Token's shouldn't be coupled in the first place since they are representing different things, one a path and other a string

wilzbach replied:
> I believe even though this ends up creating a lot of copies they should be done since essentially those Token's shouldn't be coupled in the first place since they are representing different things, one a path and other a string

Well, ideally Tokens should be immutable to avoid this from happening (https://github.com/storyscript/storyscript/issues/1056).
Anyhow, the point here was that these tokens are used from (a) the `__tmp = <expr>` insertion of the object and (b) from the `{a, b}` of the original line. Here we're talking about (b) which are unique. Think of it as e.g. moving by value as we effectively remove their prior occurrence in the Tree (though of course (a) needs to be copied).

> Thing is both of the token may look the same but need to be separate since if we move one of them into a different line and update its line number (mark_line does this)

See my response in the global thread regarding this, In short: we should make the copy when we actually overwrite the line number.

wilzbach replied:
> Here we're talking about (b) which are unique. 

Ah noticed that here's we're actually talking about `path` which is (a). Sorry about the confusion.

wilzbach commented:
> I am not so sure what are you trying to say in the second point but I am sure we are talking about the same page with the mark_lines function. I don't get whether you are suggesting to make copies elsewhere (like when I move things to other block) or are you asking why the copies are being made in the first place for somecases.

The copies are made because `mark_lines` in FakeLines overwrite the line number of their tokens. Hence, one easier solution would be simply copy there as it's actually needed there. This would also make it easier to refactor things once we move the line-parts of the FakeLine blocks after semantic (see e.g. https://github.com/storyscript/storyscript/issues/1055)

tl;dr: let's make the copies in FakeTree when they are actually needed to be copied. That should keep Lowering.py simpler and thus we can hopefully refactor is easier soon.

> Yeah some sorta pretty printing code would be awesome over there instead of just saying expression.
I thought we would add it up later on fix this but since you have mentioned it, let me study out the your draft PR

Yeah, my point was that you only need a small part of it and that part was already working back then.

adnrs96 commented:
Hey @wilzbach Thanks for the reviews and comments. I worked up on this more and cleaned up the PR a bit. It still needs work on the AST formatting thing but things should work fine for now for simple expressions which means merging this we won't be introducing regressions. I have detailed more about this in the first bullet in the notes. Wondering if that should be done before anything is merged or doing that as a separate issue makes sense?

Commit labelled `feat(parser): support expressions in list and object index` when merged is enough to solve the outstanding issue of supporting expressions in indexes.

Notes:

* For now I haven't added the changes required to deal with the `expression` regression. For simple expressions there will be no regressions and that can be seen from the reduced size of the commit changes itself. However, an input such as
```
a = {"foo": "bar"}
a[true + true]
```
will result in leakeage of internal representations as dipicated by the error message
```
Error: syntax error in story at line 2, column 1

2|    a[true + true]
      ^

E0104: `Map[string,string]` can't be indexed with `__p-2.1` of type `int`
```

I am planning to fork of another issue for this (having proper expression in error message) and deal with this there. How does that sound?
 
* Since assignment destructoring just requires a simple expression (meaning just a single string in the index) when we stop lowering everything into fake lines and allow simple expressions to be as it is (after some AST modifications of course), the need to fiddle with the assignment destructoring is no longer present.
* Regardless of the fact in the previous bullet, I still added the commits dealing with the deviation in behaviour for assignment destructoring lowering phase from the others since I believe it should be done regardless.
* I did not proceed with any work on the copying of Token's since the suggested way of doing that was "make the copy when required" and the requirement has now vanished since we changed approach and allowed simple expressions to be as it is.


Apart from this I actually discovered a few issues which I would like to detail over here:

* We are still using up line numbers for comments, though the Tree doesn't has the the comments. They are definitely ignored by the parser + grammar but line numbers are I guess determined the good old way of using `splitlines`. I am not sure if this creates an issue for the platform. I mean I don't think it should since the JSON has the correct `next` param setup but still wanted to gain more insight so brought it up.

* The logic for determining end_column is IMO not great since it isn't always doing what would be expected. I believe since it is convoluted with the error's codepath for determining what part to highlight the error is located in, things aren't well right from my persepective. Definitely needs discussion I guess.

wilzbach commented on storyscript/compiler/lowering/Faketree.py:
```diff
@@ -20,9 +20,9 @@ def __init__(self, block):
-                if tok.startswith(self.prefix):
-                    self.new_lines[tok] = False
+                tok = child.path.find_first_token()
+                if isinstance(tok.line, str) and '.' in tok.line:
```
Maybe add a comment for future readers that only fake lines have a `.` in their line number?

wilzbach commented on storyscript/compiler/lowering/Faketree.py:
```diff
@@ -78,13 +78,18 @@ def assignment(self, value):
+            # We accept and use the equals token from original expression
+            # to copy over the token meta data which helps with error messages.
+            equals.column = eq_tok.column
+            equals.end_column = eq_tok.end_column
```
Consider:

```
{a, b} = <expr>
```

-> 

```
__temp = <expr>
a = __temp["a"] # L1.2
b = __temp["b"] # L1.3
```

In other words we call `assignment_path` twice with the same `eq_tok` and now both L1.2 and L1.3 have an equal with the same start and end column. If you now look at the JSON output, you see the following:
- `a`  (`col_start`: 3, `col_end`: 11) <- here the 11 is new
- `b`  (`col_start`: 6, `col_end`: 11)

However, if you think about the potential error message (or mapping on the stack trace), it would result in sth. like `a, b } =` for `a`.

tl;dr: the status quo is broken, but this doesn't fix it either. Imho `end_column` should probably be `5` (i.e. not include `b`).

I suggest to remove this and open an issue, s.t. we can fix it correctly separately.

adnrs96 replied:
so thing is before this only the recycled node used to have the col_end and since that node used to be the last node (ie. `b`) things didn't felt that weird to include an `=` in the range of expression we were dipcting. Now since we don't recycle nodes I figured we need to make sure that the equal token has the meta data from the original equal token. This will help in isolating equals token in the original expression.

All this is related to error generation. Like if we needed to generate an error over fake line we will want to show the original line and point something out in it. That's why all fake lines having the same equals token meta data makes sense.

What's broken here is the `end_col` generation logic. I believe we should be much more granular in generating that kinda data. And the error generation code should somehow specify which all tokens to get the ranges of col, col_end for so that errors could be pointed out.

PS: If I drop the equals commit over here, I would end up losing the col_end data for the recycled nodes, which was present before in the JSON.

wilzbach replied:
> That's why all fake lines having the same equals token meta data makes sense.

Hmm, fair point. Let's go with this then :/

> nd the error generation code should somehow specify which all tokens to get the ranges of col, col_end for so that errors could be pointed out.

What do you mean? We just try to find the start of the first and end column of the last token for the erroneous AST node in question. Why would we need to look at the tokens in between?

wilzbach commented on storyscript/compiler/lowering/Lowering.py:
```diff
@@ -459,18 +460,14 @@ def visit_assignment(self, node, block, parent):
         else:
             for c in node.children:
-                self.visit_assignment(c, block, parent=node)
+                did_do_insert = self.visit_assignment(c, block, parent=node)
```
Maybe a more descriptive name like?

```suggestion
                performed_destructuring = self.visit_assignment(c, block, parent=node)
```

wilzbach commented on storyscript/compiler/lowering/Lowering.py:
```diff
@@ -584,6 +581,56 @@ def visit_function_dot(self, node, block):
+                    # lines in some common use cases.
+                    path_or_values = expression.child(0).child(0)
+                    if path_or_values.data == 'values':
+                        path_or_values = path_or_values.child(0)
```
else `assert path_or_values.data == 'path'`.

These assertions will help you a lot when changing the grammar.

wilzbach commented:
> will result in leakeage of internal representations 
> ...
> I am planning to fork of another issue for this (having proper expression in error message) and deal with this there. How does that sound?

Not great, but yeah I understand that this will make things easier for you and we can do that. We should, however, open an issue about this and don't forget to address it before 0.24.

> I still added the commits dealing with the deviation in behaviour for assignment destructoring lowering phase from the others since I believe it should be done regardless.

Yes, I agree.

> We are still using up line numbers for comments, though the Tree doesn't has the the comments. They are definitely ignored by the parser + grammar but line numbers are I guess determined the good old way of using splitlines. I am not sure if this creates an issue for the platform. I mean I don't think it should since the JSON has the correct next param setup but still wanted to gain more insight so brought it up.

That's expected as you want the line number to refer to the actual source line. The engine doesn't care about the line numbers for execution as long as they are unique. We could do `abcXXX` for all they care. tl;dr: they treat it as string label.

> The logic for determining end_column is IMO not great since it isn't always doing what would be expected. I believe since it is convoluted with the error's codepath for determining what part to highlight the error is located in, things aren't well right from my persepective. Definitely needs discussion I guess.

Can you open an issue with a concrete example?

pep8speaks commented:
Hello @adnrs96! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:



* In the file [`storyscript/compiler/lowering/Lowering.py`](https://github.com/storyscript/storyscript/blob/939037572d17044ce9bb4240ad7f079320d0f5ac/storyscript/compiler/lowering/Lowering.py):

> [Line 468:80](https://github.com/storyscript/storyscript/blob/939037572d17044ce9bb4240ad7f079320d0f5ac/storyscript/compiler/lowering/Lowering.py#L468): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (86 > 79 characters)
